### PR TITLE
Added cassandra plugin to access metrics using jolokia and push them …

### PR DESCRIPTION
**Pull request by subhachandrachandra**
This plugin is modeled after the jolokia plugin but has custom parsing logic for the json reply that is specific to cassandra. The configuration is bit simpler as it extracts metric names from the json reply. It also uses multiple metric names like "cassandraClient", "cassandraTable", "javaGarbageCollector" like the redis plugin instead of adding everything under one metric.

It has been tested against Cassandra3.x.

### CODE REVIEW

**Comment by subhachandrachandra** [comment]
Not sure what I need to do here. CircleCI failed in a kafka test.

ok  	github.com/influxdata/telegraf/plugins/outputs/graphite	1.144s
ok  	github.com/influxdata/telegraf/plugins/outputs/influxdb	1.029s
--- FAIL: TestConnectAndWrite (0.80s)
	<autogenerated>:31:                         
	Error Trace:	kafka_test.go:30		
	Error:		Received unexpected error "FAILED to send kafka message: kafka server: In the middle of a leadership election, there is currently no leader for this partition and hence it is unavailable for writes.\n"
FAIL
FAIL	github.com/influxdata/telegraf/plugins/outputs/kafka	0.823s

File: plugins/inputs/cassandra/README.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+See: https://jolokia.org/
+
+# Measurements:
+Cassandra plugin produces one or more measurements for each metric configured, adding Server's name  as `host` tag. More than one measurement is generated when querying table metrics with a wildcard for the keyspace or table name.
+
+Given a configuration like:
+
+```ini
+[cassandra]
```
Comment (sparrc) [inline plugins/inputs/cassandra/README.md]: this is not a valid telegraf configuration anymore, checkout a more recent version of the config file, it should look more like this:

```
[[inputs.cassandra]]
  ...
```

**Comment by sparrc** [comment]
servers & metrics should just be regular lists, so that configuration will look like this:

```
[[inputs.cassandra]]
  servers = [...]
  metrics = ["/heap..." ... ]
```

**Comment by subhachandrachandra** [comment]
The metrics format has been bothering me for a while. Will fix that. "servers" will be a bit tricky to add user names and passwords. Will have to figure out a format for it.

**Comment by sparrc** [comment]
in that case servers may need to be as it is, that's fine

**Comment by sparrc** [comment]
@subhachandrachandra why not use the jolokia plugin directly?

**Comment by subhachandrachandra** [comment]
@sparcc I did start with the jolokia plugin as described at https://github.com/chrusty/telegraf-cassandra.
I hit the same problems as mentioned by @chrusty. The jolokia plugin works well with the jvm metrics but  does not parse the nested json maps returned by cassandra metrics properly when the query has wild cards for keyspaces or table names. Since I had to add logic that is specific to Cassandra, I though a cassandra specific plugin will work better.

**Comment by sparrc** [comment]
okay, carry on :+1:

**Comment by subhachandrachandra** [comment]
Simplified the config to accept both servers and metrics as simple lists of strings.

File: plugins/inputs/cassandra/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+The Cassandra plugin collects Cassandra/JVM metrics exposed as MBean's attributes through jolokia REST endpoint. All metrics are collected for each server configured.
+
+See: https://jolokia.org/
+
+# Measurements:
+Cassandra plugin produces one or more measurements for each metric configured, adding Server's name  as `host` tag. More than one measurement is generated when querying table metrics with a wildcard for the keyspace or table name.
+
+Given a configuration like:
+
+```ini
```
Comment (sparrc) [inline plugins/inputs/cassandra/README.md]: should be `toml` not `ini`

File: plugins/inputs/cassandra/README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+[inputs.cassandra]
```
Comment (sparrc) [inline plugins/inputs/cassandra/README.md]: this should be `[[inputs.cassandra]]`

File: plugins/inputs/cassandra/README.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  context = "/jolokia/read"
+  servers = [":8778"]
+  metrics = ["/java.lang:type=Memory/HeapMemoryUsage"]
```
Comment (sparrc) [inline plugins/inputs/cassandra/README.md]: could you provide more examples of useful/common metrics?

File: plugins/inputs/cassandra/README.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# Telegraf plugin: Cassandra
+
+#### Plugin arguments:
+- **context** string: Context root used for jolokia url
+- **servers** []string: List of servers with the format "<user:passwd@><host>:port"
+- **metrics** []string: List of Jmx paths that identify mbeans attributes
+
+#### Description
+
+The Cassandra plugin collects Cassandra/JVM metrics exposed as MBean's attributes through jolokia REST endpoint. All metrics are collected for each server configured.
```
Comment (sparrc) [inline plugins/inputs/cassandra/README.md]: Can you link to any Cassandra documentation on this?

File: plugins/inputs/cassandra/cassandra.go [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+  context = "/jolokia/read"
+  ## List of cassandra servers exposing jolokia read service
+  servers = ["myuser:mypassword@10.10.10.1:8778","10.10.10.2:8778",":8778"]
+  ## List of metrics collected on above servers
+  ## Each metric consists of a jmx path.
+  ## This will collect all heap memory usage metrics from the jvm and 
+  ## ReadLatency metrics for all keyspaces and tables.
+  ## "type=Table" in the query works with Cassandra3.0. Older versions might
+  ## need to use "type=ColumnFamily"
+  metrics  = ["/java.lang:type=Memory/HeapMemoryUsage",
```
Comment (sparrc) [inline plugins/inputs/cassandra/cassandra.go]: for line-length, change this to:

```toml
metrics  = [
  "/java.lang:type=Memory/HeapMemoryUsage",
  "/org.apache.cassandra.metrics:type=Table,keyspace=*,scope=*,name=ReadLatency"
]
```

**Comment by sparrc** [comment]
thanks @subhachandrachandra, just a few comments to address, and can you also update the README and CHANGELOG files?

**Comment by subhachandrachandra** [comment]
Thanks for your comments @sparrc . Addressed all of them and updated README and CHANGELOG. Added a Thank You to myself in the CHANGELOG so you don't have to edit the file again :-)

**Comment by sparrc** [comment]
@subhachandrachandra can you rebase your changes off master one more time?

**Comment by sparrc** [comment]
@subhachandrachandra your branch and/or fork appears to have diverged a few times from the influxdata master branch. Could you please squash your changes into one commit and rebase them on top of master?

**Comment by subhachandrachandra** [comment]
Closing this request. Will create new one after squashing my changes.

**Pull request closed without merging**